File: mobile/android/TrainBlink/app/src/main/java/org/trainblink/app/models/Trip.kt
package org.trainblink.app.models

import android.os.Parcelable
import com.google.gson.annotations.SerializedName
import kotlinx.parcelize.Parcelize
import java.util.Date
import java.util.UUID

/**
 * Trip status enumeration
 */
enum class TripStatus {
    @SerializedName("active")
    ACTIVE,

    @SerializedName("ended")
    ENDED,

    @SerializedName("cancelled")
    CANCELLED
}

/**
 * Trip model representing a user's journey
 */
@Parcelize
data class Trip(
    val id: UUID,
    @SerializedName("user_id")
    val userId: UUID,
    val route: String,
    @SerializedName("train_number")
    val trainNumber: String?,
    @SerializedName("departure_time")
    val departureTime: Date,
    @SerializedName("estimated_arrival")
    val estimatedArrival: Date,
    @SerializedName("actual_end_time")
    val actualEndTime: Date?,
    @SerializedName("discovery_enabled")
    val discoveryEnabled: Boolean,
    @SerializedName("ble_anonymous_id")
    val bleAnonymousId: String,
    val status: TripStatus,
    @SerializedName("created_at")
    val createdAt: Date,
    @SerializedName("updated_at")
    val updatedAt: Date
) : Parcelable {

    /**
     * Check if trip is currently active
     */
    val isActive: Boolean
        get() = status == TripStatus.ACTIVE && Date().before(estimatedArrival)

    /**
     * Calculate trip duration in milliseconds
     */
    val duration: Long
        get() = estimatedArrival.time - departureTime.time

    /**
     * Calculate remaining time in milliseconds
     */
    val remainingTime: Long
        get() = maxOf(0, estimatedArrival.time - Date().time)

    /**
     * Format remaining time as string
     */
    val remainingTimeFormatted: String
        get() {
            val remaining = remainingTime
            val hours = (remaining / (1000 * 60 * 60)).toInt()
            val minutes = ((remaining / (1000 * 60)) % 60).toInt()

            return if (hours > 0) {
                "${hours}h ${minutes}m remaining"
            } else {
                "${minutes}m remaining"
            }
        }
}

/**
 * Request payload for starting a new trip
 */
data class StartTripRequest(
    val route: String,
    @SerializedName("train_number")
    val trainNumber: String?,
    @SerializedName("departure_time")
    val departureTime: Date,
    @SerializedName("estimated_arrival")
    val estimatedArrival: Date
)

/**
 * Request payload for ending a trip
 */
data class EndTripRequest(
    @SerializedName("trip_id")
    val tripId: UUID
)

/**
 * API response wrapper for single trip
 */
data class TripResponse(
    val status: String,
    val data: Trip?,
    val error: String?
)

/**
 * API response wrapper for trip list
 */
data class TripsListResponse(
    val status: String,
    val data: TripsData?,
    val error: String?
) {
    data class TripsData(
        val trips: List<Trip>,
        val total: Int,
        val limit: Int,
        val offset: Int
    )
}

/**
 * Update discovery enabled request
 */
data class UpdateDiscoveryRequest(
    @SerializedName("discovery_enabled")
    val discoveryEnabled: Boolean
)

File: mobile/ios/TrainBlink/TrainBlink/Models/Trip.swift
//
//  Trip.swift
//  TrainBlink
//
//  Core trip model for tracking user journeys
//

import Foundation

/// Trip status enumeration
enum TripStatus: String, Codable {
    case active = "active"
    case ended = "ended"
    case cancelled = "cancelled"
}

/// Trip model representing a user's journey
struct Trip: Codable, Identifiable {
    let id: UUID
    let userId: UUID
    let route: String
    let trainNumber: String?
    let departureTime: Date
    let estimatedArrival: Date
    var actualEndTime: Date?
    var discoveryEnabled: Bool
    let bleAnonymousId: String
    var status: TripStatus
    let createdAt: Date
    let updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case route
        case trainNumber = "train_number"
        case departureTime = "departure_time"
        case estimatedArrival = "estimated_arrival"
        case actualEndTime = "actual_end_time"
        case discoveryEnabled = "discovery_enabled"
        case bleAnonymousId = "ble_anonymous_id"
        case status
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }

    /// Check if trip is currently active
    var isActive: Bool {
        return status == .active && Date() < estimatedArrival
    }

    /// Calculate trip duration
    var duration: TimeInterval {
        return estimatedArrival.timeIntervalSince(departureTime)
    }

    /// Calculate remaining time
    var remainingTime: TimeInterval {
        return max(0, estimatedArrival.timeIntervalSince(Date()))
    }

    /// Format remaining time as string
    var remainingTimeFormatted: String {
        let remaining = remainingTime
        let hours = Int(remaining) / 3600
        let minutes = Int(remaining) % 3600 / 60

        if hours > 0 {
            return "\(hours)h \(minutes)m remaining"
        } else {
            return "\(minutes)m remaining"
        }
    }
}

/// Request payload for starting a new trip
struct StartTripRequest: Codable {
    let route: String
    let trainNumber: String?
    let departureTime: Date
    let estimatedArrival: Date

    enum CodingKeys: String, CodingKey {
        case route
        case trainNumber = "train_number"
        case departureTime = "departure_time"
        case estimatedArrival = "estimated_arrival"
    }
}

/// Request payload for ending a trip
struct EndTripRequest: Codable {
    let tripId: UUID

    enum CodingKeys: String, CodingKey {
        case tripId = "trip_id"
    }
}

/// API response wrapper
struct TripResponse: Codable {
    let status: String
    let data: Trip?
    let error: String?
}

struct TripsListResponse: Codable {
    let status: String
    let data: TripsData?
    let error: String?

    struct TripsData: Codable {
        let trips: [Trip]
        let total: Int
        let limit: Int
        let offset: Int
    }
}
